File: internal/logic/approval.go
package logic

import (
	"context"
	"fmt"

	"aiOffice/internal/domain"
	"aiOffice/internal/model"
	"aiOffice/internal/svc"
	"aiOffice/pkg/timeutils"
	"aiOffice/pkg/xerr"
)

var (
	ErrApprovalNotFound = fmt.Errorf("审批不存在")
)

type Approval interface {
	Info(ctx context.Context, req *domain.IdPathReq) (resp *domain.ApprovalInfoResp, err error)
	Create(ctx context.Context, req *domain.Approval) (resp *domain.IdResp, err error)
	Dispose(ctx context.Context, req *domain.DisposeReq) (err error)
	List(ctx context.Context, req *domain.ApprovalListReq) (resp *domain.ApprovalListResp, err error)
}

type approval struct {
	svcCtx *svc.ServiceContext
}

func NewApproval(svcCtx *svc.ServiceContext) Approval {
	return &approval{
		svcCtx: svcCtx,
	}
}

// Info 获取审批详情
func (l *approval) Info(ctx context.Context, req *domain.IdPathReq) (resp *domain.ApprovalInfoResp, err error) {
	approvalData, err := l.svcCtx.ApprovalModel.FindOne(ctx, req.Id)
	if err != nil {
		if err == model.ErrNotFound {
			return nil, ErrApprovalNotFound
		}
		return nil, xerr.WithMessage(err, "查询审批失败")
	}

	// 使用model中的转换方法
	resp = approvalData.ToDomainApprovalInfo()

	// 获取申请人信息
	user, err := l.svcCtx.UserModel.FindOne(ctx, approvalData.UserId)
	if err == nil {
		resp.User = &domain.Approver{
			UserId:   user.ID.Hex(),
			UserName: user.Name,
		}
	}

	// 转换审批人列表
	for _, a := range approvalData.Approvers {
		resp.Approvers = append(resp.Approvers, &domain.Approver{
			UserId:   a.UserId,
			UserName: a.UserName,
			Status:   int(a.Status),
			Reason:   a.Reason,
		})
	}

	// 转换抄送人列表
	for _, c := range approvalData.CopyPersons {
		resp.CopyPersons = append(resp.CopyPersons, &domain.Approver{
			UserId:   c.UserId,
			UserName: c.UserName,
			Status:   int(c.Status),
		})
	}

	// 设置当前审批人
	if approvalData.ApprovalIdx < len(approvalData.Approvers) {
		currentApprover := approvalData.Approvers[approvalData.ApprovalIdx]
		resp.Approver = &domain.Approver{
			UserId:   currentApprover.UserId,
			UserName: currentApprover.UserName,
			Status:   int(currentApprover.Status),
			Reason:   currentApprover.Reason,
		}
	}

	return resp, nil
}

// Create 创建审批
func (l *approval) Create(ctx context.Context, req *domain.Approval) (resp *domain.IdResp, err error) {
	// 生成审批编号
	no := fmt.Sprintf("SP%d", timeutils.Now())

	approvalData := &model.Approval{
		UserId:   req.UserId,
		No:       no,
		Type:     model.ApprovalType(req.Type),
		Status:   model.Processed, // 初始状态：处理中
		Title:    req.Title,
		Abstract: req.Abstract,
		Reason:   req.Reason,
	}

	// 根据审批类型设置详情
	switch model.ApprovalType(req.Type) {
	case model.LeaveApproval:
		if req.Leave != nil {
			approvalData.Leave = &model.Leave{
				Type:      model.LeaveType(req.Leave.Type),
				StartTime: req.Leave.StartTime,
				EndTime:   req.Leave.EndTime,
				Reason:    req.Leave.Reason,
				TimeType:  model.TimeFormatType(req.Leave.TimeType),
			}
			approvalData.Title = model.LeaveType(req.Leave.Type).ToString()
		}
	case model.MakeCardApproval:
		if req.MakeCard != nil {
			approvalData.MakeCard = &model.MakeCard{
				Date:      req.MakeCard.Date,
				Reason:    req.MakeCard.Reason,
				Day:       req.MakeCard.Day,
				CheckType: model.WorkCheckType(req.MakeCard.CheckType),
			}
			approvalData.Title = model.ApprovalType(req.Type).ToString()
		}
	case model.GoOutApproval:
		if req.GoOut != nil {
			approvalData.GoOut = &model.GoOut{
				StartTime: req.GoOut.StartTime,
				EndTime:   req.GoOut.EndTime,
				Duration:  req.GoOut.Duration,
				Reason:    req.GoOut.Reason,
			}
			approvalData.Title = model.ApprovalType(req.Type).ToString()
		}
	default:
		approvalData.Title = model.ApprovalType(req.Type).ToString()
	}

	err = l.svcCtx.ApprovalModel.Insert(ctx, approvalData)
	if err != nil {
		return nil, xerr.WithMessage(err, "创建审批失败")
	}

	return &domain.IdResp{Id: approvalData.ID.Hex()}, nil
}

// Dispose 处理审批（通过/拒绝）
func (l *approval) Dispose(ctx context.Context, req *domain.DisposeReq) (err error) {
	approvalData, err := l.svcCtx.ApprovalModel.FindOne(ctx, req.ApprovalId)
	if err != nil {
		if err == model.ErrNotFound {
			return ErrApprovalNotFound
		}
		return xerr.WithMessage(err, "查询审批失败")
	}

	// 检查审批状态是否为处理中
	if approvalData.Status != model.Processed {
		return xerr.New(fmt.Errorf("审批已处理"))
	}

	// 更新当前审批人的状态
	if approvalData.ApprovalIdx < len(approvalData.Approvers) {
		approvalData.Approvers[approvalData.ApprovalIdx].Status = model.ApprovalStatus(req.Status)
		approvalData.Approvers[approvalData.ApprovalIdx].Reason = req.Reason
	}

	// 根据处理结果更新审批状态
	switch model.ApprovalStatus(req.Status) {
	case model.Pass:
		// 检查是否还有下一个审批人
		if approvalData.ApprovalIdx+1 < len(approvalData.Approvers) {
			// 移动到下一个审批人
			approvalData.ApprovalIdx++
			approvalData.ApprovalId = approvalData.Approvers[approvalData.ApprovalIdx].UserId
		} else {
			// 所有审批人都通过，审批完成
			approvalData.Status = model.Pass
			approvalData.FinishAt, approvalData.FinishDay, approvalData.FinishMonth, approvalData.FinishYeas = timeutils.FinishTime()
		}
	case model.Refuse:
		// 拒绝，审批结束
		approvalData.Status = model.Refuse
		approvalData.FinishAt, approvalData.FinishDay, approvalData.FinishMonth, approvalData.FinishYeas = timeutils.FinishTime()
	}

	err = l.svcCtx.ApprovalModel.Update(ctx, approvalData)
	if err != nil {
		return xerr.WithMessage(err, "更新审批失败")
	}

	return nil
}

// List 审批列表
func (l *approval) List(ctx context.Context, req *domain.ApprovalListReq) (resp *domain.ApprovalListResp, err error) {
	approvals, total, err := l.svcCtx.ApprovalModel.List(ctx, req.UserId, req.Type, req.Page, req.Count)
	if err != nil {
		return nil, xerr.WithMessage(err, "查询审批列表失败")
	}

	resp = &domain.ApprovalListResp{
		Count: total,
		List:  make([]*domain.ApprovalList, 0, len(approvals)),
	}

	for _, a := range approvals {
		resp.List = append(resp.List, a.ToDomainApprovalList())
	}

	return resp, nil
}

File: aioffice-front/src/types/approval.ts
/**
 * Approval type definitions
 * Requirements: 5.1 - Approval management
 */

// 审批人信息
export interface Approver {
  userId: string;
  userName: string;
  status: number;
  reason?: string;
}

export interface Approval {
  id: string;
  userId: string;
  no: string;
  type: number; // 1=事假, 2=调休, 3=病假, 4=年假, 5=产假, 6=陪产假, 7=婚假, 8=丧假, 9=哺乳假
  status: number; // 0=未开始, 1=进行中, 2=已通过, 3=已撤销, 4=已拒绝
  title: string;
  abstract: string;
  reason: string;
  // 详情接口返回的额外字段
  user?: Approver;
  approver?: Approver;
  approvers?: Approver[];
  copyPersons?: Approver[];
  // 完成时间相关
  finishAt?: number;
  finishDay?: number;
  finishMonth?: number;
  finishYeas?: number;
  // 具体审批类型数据
  leave?: LeaveInfo;
  goOut?: GoOutInfo;
  makeCard?: MakeCardInfo;
  createAt: number;
  updateAt: number;
}

export interface LeaveInfo {
  type: number;
  startTime: number;
  endTime: number;
  duration: number;
  reason: string;
  timeType: number; // 1=小时, 2=天
}

export interface GoOutInfo {
  startTime: number;
  endTime: number;
  duration: number;
  reason: string;
}

export interface MakeCardInfo {
  date?: number;           // 补卡时间
  reason?: string;         // 补卡理由
  day?: number;            // 补卡日期(20221011)
  workCheckType?: number;  // 补卡类型
}

export interface ApprovalState {
  approvals: Approval[];
  loading: boolean;
  currentApproval: Approval | null;
}

// 创建审批的基础字段
export interface CreateApprovalBase {
  userId: string;       // 用户ID
  type: number;         // 审批类型
  title?: string;       // 标题
  abstract?: string;    // 摘要
  reason?: string;      // 原因
}

// 创建请假审批
export interface CreateLeaveApprovalRequest extends CreateApprovalBase {
  leave: LeaveInfo;
}

// 创建外出审批
export interface CreateGoOutApprovalRequest extends CreateApprovalBase {
  goOut: GoOutInfo;
}

// 创建补卡审批
export interface CreateMakeCardApprovalRequest extends CreateApprovalBase {
  makeCard: MakeCardInfo;
}

export interface DisposeApprovalRequest {
  status: number;      // 审批状态
  reason: string;      // 审批理由
  approvalId: string;  // 审批ID
}

// Approval type constants
export const ApprovalType = {
  PERSONAL_LEAVE: 1,  // 事假
  COMPENSATORY: 2,    // 调休
  SICK_LEAVE: 3,      // 病假
  ANNUAL_LEAVE: 4,    // 年假
  MATERNITY: 5,       // 产假
  PATERNITY: 6,       // 陪产假
  MARRIAGE: 7,        // 婚假
  BEREAVEMENT: 8,     // 丧假
  NURSING: 9,         // 哺乳假
} as const;

// Approval status constants
export const ApprovalStatus = {
  NOT_STARTED: 0,     // 未开始
  IN_PROGRESS: 1,     // 进行中
  APPROVED: 2,        // 已通过
  CANCELLED: 3,       // 已撤销
  REJECTED: 4,        // 已拒绝
} as const;
